Array transform, array filter, and array reduce function

haozhun opened the pull request.

cberner commented on presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java:
```diff
@@ -198,13 +198,15 @@
 import static com.facebook.presto.operator.aggregation.MinNAggregationFunction.MIN_N_AGGREGATION;
 import static com.facebook.presto.operator.aggregation.MultimapAggregationFunction.MULTIMAP_AGG;
 import static com.facebook.presto.operator.scalar.ArrayConstructor.ARRAY_CONSTRUCTOR;
+import static com.facebook.presto.operator.scalar.ArrayFilterFunction.ARRAY_FILTER_FUNCTION;
```
Seems like this is part of the wrong commit

haozhun replied:
Fixed

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayTransformFunction.java:
```diff
@@ -0,0 +1,104 @@
+    @Override
+    public boolean isHidden()
+    {
+        return true;
```
Why is this hidden?

haozhun replied:
Fixed

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFilterFunction.java:
```diff
@@ -0,0 +1,106 @@
+    @Override
+    public boolean isHidden()
+    {
+        return true;
```
Why hidden?

haozhun replied:
Fixed

cberner commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java:
```diff
@@ -362,14 +362,16 @@ protected Type visitDereferenceExpression(DereferenceExpression node, StackableA
-                }
-                if (!scope.isColumnReference(qualifiedName)) {
-                    throwMissingAttributeException(node, qualifiedName);
+            if (!context.getContext().isInLambda()) {
```
Add a test?

haozhun replied:
done

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayReduceFunction.java:
```diff
@@ -0,0 +1,120 @@
+                ImmutableList.of(typeVariable("T"), typeVariable("S"), typeVariable("R")),
+                ImmutableList.of(),
+                parseTypeSignature("R"),
+                ImmutableList.of(parseTypeSignature("array(T)"), parseTypeSignature("S"), parseTypeSignature("function(T,S,S)"), parseTypeSignature("function(S,R)")),
```
Maybe we should also have a form that's `reduce(array(T), R, function(T, R, R))`? Since it's probably common to have an identity transformation from the intermediate state to the final state

haozhun replied:
I'm not sure about this.

Benefit of having a 3-argument version:
* User don't need to put `, s -> s` at the end of their `array_reduce` call.

Benefit of not having a 3-argument version:
* All user (instead of few users) will know they can implement something more complicated (like average) without writing two or more reduce functions.

cberner replied:
Ok, lets add it later if people complain

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayReduceFunction.java:
```diff
@@ -0,0 +1,120 @@
+    @Override
+    public boolean isHidden()
+    {
+        return true;
```
`false`?

haozhun replied:
Fixed

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayFilterFunction.java:
```diff
@@ -0,0 +1,106 @@
+            catch (Throwable throwable) {
+                throw Throwables.propagate(throwable);
+            }
+            if (TRUE.equals(keep)) {
```
I'm not sure this has the right semantics. Since `NULL` means "unknown" if the result of the function is `NULL` shouldn't the output for that position be `NULL` instead of omitting it?

haozhun replied:
When you do
```
SELECT *
FROM (VALUES 1, 2, null, 4, null, 5) t(x)
WHERE x > 0
```
You get 1, 2, 4, 5 instead of 1, 2, null, 4, null, 5.

This is the reason why I skip "unknown"s.

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayReduceFunction.java:
```diff
@@ -0,0 +1,120 @@
+    {
+        int positionCount = block.getPositionCount();
+        if (positionCount == 0) {
+            return null;
```
Why `null`? It seems like it should use the initial intermediate value in this case, or is this how other languages handle it?

haozhun replied:
@dain and I had a discussion on this after I talked about this behavior.

The choice was made hear to mimic SQL aggregation function behavior. I made a mistake in the original code. I shouldn't treat zero-length array specially. I should instead treat null-only input array (including empty array) specially. I fixed it.

For users who need full control on null element in the input array, the decision we made at the time was that we could introduce a `array_reduce_nullable`.

The choice was also made because the difference in easiness to go from one way to another

* If the defined behavior matches user expection:
  * `array_reduce(arr, initVal, (x, s) -> f(x, s))`
* Otherwise
  * If array_reduce produces null for empty, to produce non-null for null-only input (including empty):
    * `coalesce(array_reduce(arr, initVal, (x, s) -> f(x, s)), initVal)`
  * If array_reduce produces initial value for empty, to produce null for null-only input (including empty):
    * Assuming `f` never produces null:
      * `array_reduce(arr, null, (x, s) -> IF(s IS NULL, f(x, initVal), f(x, s)))`
    * Otherwise:

```
       array_reduce(
         arr,
         CAST(ROW(initVal, false) AS ROW(T val, BOOLEAN hasNonNull)),
         (x, s) -> IF(
           x IS NULL,
           s,
           CAST(ROW(f(x, s), true) AS ROW(T val, BOOLEAN hasNonNull))),
         s -> IF(s.hasNonNull, s.val, null))
```

cberner commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayReduceFunction.java:
```diff
@@ -0,0 +1,120 @@
+        MethodHandle methodHandle = METHOD_HANDLE.bindTo(inputType).bindTo(intermediateType).bindTo(outputType);
+        return new ScalarFunctionImplementation(
+                true,
+                ImmutableList.of(false, false, false, false),
```
Seems like you should allow `NULL` for the second argument, since a `NULL` initial state seems reasonable

cberner replied:
A concrete use-case would be implementing a "max" reduction. To make it work like the `max` aggregation it should return `NULL` for empty arrays. It seems like the natural way to do that would be `reduce(my_array, NULL, (x, s) -> greatest(x, coalesce(s, x)), s -> s)`

haozhun commented:
Travis (before addressing comments) passes after I restarted it.

I talked with @martint about the names. We agreed on using `transform`, `filter`, `reduce`.

cberner commented:
I think you missed my comment about benchmarking `invokeExact` vs `invoke`

haozhun commented:
The benchmark does show quite a penalty for using invoke

Test scenario:
* BIGINT
* cheap filter (>0)
* 4 items per array (random)

```
Benchmark                             (name)  Mode  Cnt    Score    Error  Units
BenchmarkArrayFilter.benchmark        filter  avgt   20  143.856 ±  8.426  ns/op
BenchmarkArrayFilter.benchmark        filter  avgt   20  154.264 ± 15.690  ns/op
BenchmarkArrayFilter.benchmark  exact_filter  avgt   20  103.174 ±  9.264  ns/op
BenchmarkArrayFilter.benchmark  exact_filter  avgt   20   93.488 ±  5.584  ns/op
```

However, I don't see how I can improve based on this observation unless:
* writing 5 implementations for filter
* byte code generation or writing 25 implementations for transform
* byte code generation or writing 125 implementations for reduce

I feel like such improvements are quite complex and can wait.

I also tried casting the method handle at the beginning of the function (using a pre-bound MethodType), it doesn't help.
```
Benchmark                             (name)  Mode  Cnt    Score    Error  Units
BenchmarkArrayFilter.benchmark        filter  avgt   20  152.011 ± 14.979  ns/op
BenchmarkArrayFilter.benchmark  exact_filter  avgt   20  160.646 ± 13.269  ns/op
```

martint commented on presto-main/src/test/java/com/facebook/presto/operator/scalar/TestArrayReduceFunction.java:
```diff
@@ -0,0 +1,91 @@
+    {
+        assertFunction("reduce(ARRAY [NULL], CAST (0 AS BIGINT), (x, s) -> x + s, s -> s)", BIGINT, null);
+        assertFunction("reduce(ARRAY [NULL, NULL, NULL], CAST (0 AS BIGINT), (x, s) -> x + s, s -> s)", BIGINT, null);
+        assertFunction("reduce(ARRAY [5, NULL, 50], CAST (0 AS BIGINT), (x, s) -> x + s, s -> s)", BIGINT, 55L);
```
The semantics of reduce seem weird and inconsistent. In particular, I find it odd that:

`reduce(array[null], 1, (x, s) -> x + s, s -> s)` => `null`

but 

`reduce(array[null, 0], 1, (x, s) -> x + s, s -> s)` => `1`

In the first case, it's understandable why it would produce null, since it's equivalent to `1 + null`.

In the second case, it's should produce the same as `1 + null + 0`, which means it should also produce null if we're to be consistent.

martint replied:
And `reduce(array[], 1, (x, s) -> x + s, s -> s)` should produce `1`, not `null`, by the same argument.

haozhun replied:
See [this comment](https://github.com/prestodb/presto/pull/6721#discussion_r90311832. In short, this mimics how SQL aggregation function works. We can discuss further since both you and @cberner raised the same concern.

martint replied:
Yes, let's discuss. SQL aggregations have quirky semantics, so I don't think we should model them that way.

cberner commented:
The implementation looks good, but we should discuss the semantics. Also, can you file a follow-up issue for the performance optimizations? That's probably a good beginner/intermediate task (depending on which function)

haozhun commented:
Updated reduce function:
* Initial state is now nullable.
* Null elements are no longer skipped.
* Empty arrays no longer forces null result.
